File: CSharp/SpaceVIL/SpaceVIL/Engine/CommonProcessor.cs
using System;
using System.Collections.Generic;
using Glfw3;
using SpaceVIL.Common;
using SpaceVIL.Core;

namespace SpaceVIL
{
    internal class CommonProcessor
    {
        internal WindowProcessor WndProcessor;
        internal ToolTipItem Tooltip;
        internal ActionManagerAssigner Manager;
        internal GLWHandler Handler;
        internal CoreWindow Window;
        internal WindowLayout Layout;
        internal WContainer RootContainer;
        internal Guid Guid;
        internal MouseArgs Margs;
        internal List<Prototype> UnderHoveredItems;
        internal List<Prototype> UnderFocusedItems;
        internal Prototype DraggableItem = null;
        internal Prototype HoveredItem = null;
        internal Prototype FocusedItem = null;
        internal int WGlobal = 0;
        internal int HGlobal = 0;
        internal int XGlobal = 0;
        internal int YGlobal = 0;
        internal bool InputLocker = false;
        internal Position PtrPress = new Position();
        internal Position PtrRelease = new Position();
        internal Position PtrClick = new Position();
        internal InputDeviceEvent Events;

        internal CommonProcessor()
        {
            this.Events = new InputDeviceEvent();
            Margs = new MouseArgs();
            Margs.Clear();
            UnderFocusedItems = new List<Prototype>();
            UnderHoveredItems = new List<Prototype>();
            WndProcessor = new WindowProcessor(this);
        }

        internal void InitProcessor(GLWHandler handler, ToolTipItem toolTip)
        {
            this.Handler = handler;
            this.Tooltip = toolTip;
            Window = handler.GetCoreWindow();
            Layout = Window.GetLayout();
            RootContainer = Layout.GetContainer();
            Guid = Window.GetWindowGuid();
            this.Manager = new ActionManagerAssigner(Layout);
        }

        internal bool GetHoverPrototype(float xpos, float ypos, InputEventType action)
        {
            InputLocker = true;
            UnderHoveredItems.Clear();

            List<IBaseItem> layout_box_of_items = new List<IBaseItem>();
            layout_box_of_items.Add(RootContainer);
            layout_box_of_items.AddRange(GetInnerItems(RootContainer, xpos, ypos));

            foreach (var item in ItemsLayoutBox.GetLayoutFloatItems(Guid))
            {
                if (!item.IsVisible() || !item.IsDrawable())
                    continue;
                layout_box_of_items.Add(item);

                Prototype leaf = item as Prototype;
                if (leaf != null)
                    layout_box_of_items.AddRange(GetInnerItems(leaf, xpos, ypos));
            }
            InputLocker = false;

            List<Prototype> queue = new List<Prototype>();

            foreach (var item in layout_box_of_items)
            {
                Prototype prototype = item as Prototype;
                if (prototype != null)
                {
                    if (prototype.GetHoverVerification(xpos, ypos))
                    {
                        queue.Add(prototype);
                    }
                    else
                    {
                        prototype.SetMouseHover(false);
                        IFloating floatItem = item as IFloating;
                        if (floatItem != null && action == InputEventType.MousePress)
                        {
                            if (floatItem.IsOutsideClickClosable())
                            {
                                ContextMenu cmToClose = (item as ContextMenu);
                                if (cmToClose != null)
                                {
                                    if (cmToClose.CloseDependencies(Margs))
                                        floatItem.Hide();
                                }
                                else
                                {
                                    floatItem.Hide(Margs);
                                }
                            }
                        }
                    }
                }
            }

            if (queue.Count > 0)
            {
                if (HoveredItem != null && (HoveredItem != queue[queue.Count - 1]))
                    Manager.AssignActionsForSender(InputEventType.MouseLeave, Margs, HoveredItem, UnderFocusedItems, false);

                HoveredItem = queue[queue.Count - 1];
                HoveredItem.SetMouseHover(true);
                CommonService.CurrentCursor = HoveredItem.GetCursor();
                Glfw.SetCursor(Handler.GetWindowId(), CommonService.CurrentCursor.GetCursor());

                if (Handler.GetCoreWindow().IsBorderHidden && Handler.GetCoreWindow().IsResizable && !Handler.GetCoreWindow().IsMaximized)
                {
                    int handlerContainerWidth = Handler.GetCoreWindow().GetLayout().GetContainer().GetWidth();
                    int handlerContainerHeight = Handler.GetCoreWindow().GetLayout().GetContainer().GetHeight();

                    bool cursorNearLeftTop = (xpos <= SpaceVILConstants.BorderCursorTolerance && ypos <= SpaceVILConstants.BorderCursorTolerance);
                    bool cursorNearLeftBottom = (ypos >= handlerContainerHeight - SpaceVILConstants.BorderCursorTolerance && xpos <= SpaceVILConstants.BorderCursorTolerance);
                    bool cursorNearRightTop = (xpos >= handlerContainerWidth - SpaceVILConstants.BorderCursorTolerance && ypos <= SpaceVILConstants.BorderCursorTolerance);
                    bool cursorNearRightBottom = (xpos >= handlerContainerWidth - SpaceVILConstants.BorderCursorTolerance && ypos >= handlerContainerHeight - SpaceVILConstants.BorderCursorTolerance);

                    if (cursorNearRightTop || cursorNearRightBottom || cursorNearLeftBottom || cursorNearLeftTop)
                    {
                        Handler.SetCursorType(EmbeddedCursor.Crosshair);
                    }
                    else
                    {
                        if (xpos >= handlerContainerWidth - SpaceVILConstants.BorderCursorTolerance || xpos < SpaceVILConstants.BorderCursorTolerance)
                            Handler.SetCursorType(EmbeddedCursor.ResizeX);

                        if (ypos >= handlerContainerHeight - SpaceVILConstants.BorderCursorTolerance || ypos < SpaceVILConstants.BorderCursorTolerance)
                            Handler.SetCursorType(EmbeddedCursor.ResizeY);
                    }
                }

                UnderHoveredItems = queue;
                Stack<Prototype> tmp = new Stack<Prototype>(UnderHoveredItems);
                while (tmp.Count > 0)
                {
                    Prototype item = tmp.Pop();
                    if (item.Equals(HoveredItem) && HoveredItem.IsDisabled())
                        continue;
                    item.SetMouseHover(true);
                    if (!item.IsPassEvents(InputEventType.MouseHover))
                        break;
                }

                Manager.AssignActionsForHoveredItem(InputEventType.MouseHover, Margs, HoveredItem, UnderHoveredItems, false);
                return true;
            }
            else
                return false;
        }

        private List<IBaseItem> GetInnerItems(Prototype root)
        {
            List<IBaseItem> list = new List<IBaseItem>();
            List<IBaseItem> root_items = root.GetItems();
            foreach (var item in root_items)
            {
                if (!item.IsVisible() || !item.IsDrawable())
                    continue;

                Prototype leaf = item as Prototype;
                if (leaf != null)
                {
                    if (leaf.IsDisabled())
                        continue;
                    list.Add(item);
                    list.AddRange(GetInnerItems(leaf));
                }
            }
            return list;
        }

        private List<IBaseItem> GetInnerItems(Prototype root, float xpos, float ypos)
        {
            List<IBaseItem> list = new List<IBaseItem>();
            List<IBaseItem> rootItems = root.GetItems();

            foreach (IBaseItem item in rootItems)
            {
                if (!item.IsVisible() || !item.IsDrawable())
                    continue;
                Prototype leaf = item as Prototype;
                if (leaf != null)
                {
                    if (leaf.IsDisabled())
                        continue;
                    if (leaf.GetHoverVerification(xpos, ypos))
                        list.Add(item);
                    list.AddRange(GetInnerItems(leaf, xpos, ypos));
                }
            }
            return list;
        }

        internal Prototype IsInListHoveredItems<T>()
        {
            Prototype wanted = null;
            foreach (var item in UnderHoveredItems)
            {
                if (item is T)
                {
                    wanted = item;
                    if (wanted is IFloating)
                        return wanted;
                }
            }
            return wanted;
        }

        internal void FindUnderFocusedItems(Prototype item)
        {
            Stack<Prototype> queue = new Stack<Prototype>();
            if (item == RootContainer)
            {
                UnderFocusedItems = null;
                return;
            }
            Prototype parent = item.GetParent();
            while (parent != null)
            {
                queue.Push(parent);
                parent = parent.GetParent();
            }
            UnderFocusedItems = new List<Prototype>(queue);
            UnderFocusedItems.Remove(FocusedItem);
        }

        internal void SetFocusedItem(Prototype item)
        {
            if (item == null)
            {
                FocusedItem = null;
                return;
            }
            if (FocusedItem != null)
            {
                if (FocusedItem.Equals(item))
                    return;
                FocusedItem.SetFocused(false);
            }
            FocusedItem = item;
            FocusedItem.SetFocused(true);
            FindUnderFocusedItems(item);
        }

        internal void ResetFocus()
        {
            if (FocusedItem != null)
                FocusedItem.SetFocused(false);
            FocusedItem = RootContainer;
            FocusedItem.SetFocused(true);
            UnderFocusedItems?.Clear();
        }

        internal void ResetItems()
        {
            ResetFocus();
            if (HoveredItem != null)
                HoveredItem.SetMouseHover(false);
            HoveredItem = null;
            UnderHoveredItems.Clear();
        }
    }
}

File: Java/SpaceVIL/spacevil/src/main/java/com/spvessel/spacevil/CommonProcessor.java
package com.spvessel.spacevil;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.util.ArrayDeque;
import java.util.Deque;
import java.util.LinkedList;
import java.util.List;
import java.util.UUID;

import com.spvessel.spacevil.internal.Wrapper.GlfwWrapper;

import com.spvessel.spacevil.Common.CommonService;
import com.spvessel.spacevil.Core.IBaseItem;
import com.spvessel.spacevil.Core.IFloating;
import com.spvessel.spacevil.Core.MouseArgs;
import com.spvessel.spacevil.Core.Position;
import com.spvessel.spacevil.Flags.EmbeddedCursor;
import com.spvessel.spacevil.Flags.InputEventType;

final class CommonProcessor {
    WindowProcessor wndProcessor;
    ToolTipItem toolTip;
    ActionManagerAssigner manager;
    GLWHandler handler;
    CoreWindow window;
    WindowLayout layout;
    WContainer rootContainer;
    UUID guid;
    MouseArgs margs;
    List<Prototype> underHoveredItems;
    List<Prototype> underFocusedItems;
    Prototype draggableItem = null;
    Prototype hoveredItem = null;
    Prototype focusedItem = null;
    int wGlobal = 0;
    int hGlobal = 0;
    int xGlobal = 0;
    int yGlobal = 0;
    boolean inputLocker = false;
    Position ptrPress = new Position();
    Position ptrRelease = new Position();
    Position ptrClick = new Position();
    InputDeviceEvent events;

    GlfwWrapper glfw = null;

    CommonProcessor() {
        glfw = GlfwWrapper.get();
        this.events = new InputDeviceEvent();
        margs = new MouseArgs();
        margs.clear();
        underFocusedItems = new LinkedList<>();
        underHoveredItems = new LinkedList<>();
        wndProcessor = new WindowProcessor(this);
    }

    void initProcessor(GLWHandler handler, ToolTipItem toolTip) {
        this.handler = handler;
        this.toolTip = toolTip;
        window = handler.getCoreWindow();
        layout = window.getLayout();
        rootContainer = layout.getContainer();
        guid = window.getWindowGuid();
        this.manager = new ActionManagerAssigner(layout);
    }

    boolean getHoverPrototype(float xpos, float ypos, InputEventType action) {
        inputLocker = true;
        underHoveredItems.clear();

        List<IBaseItem> layout_box_of_items = new LinkedList<IBaseItem>();
        layout_box_of_items.add(rootContainer);
        layout_box_of_items.addAll(getInnerItems(rootContainer, xpos, ypos));

        for (IBaseItem item : ItemsLayoutBox.getLayoutFloatItems(guid)) {
            if (!item.isVisible() || !item.isDrawable())
                continue;
            layout_box_of_items.add(item);

            if (item instanceof Prototype)
                layout_box_of_items.addAll(getInnerItems((Prototype) item, xpos, ypos));
        }
        inputLocker = false;

        List<Prototype> queue = new LinkedList<>();

        for (IBaseItem item : layout_box_of_items) {
            if (item instanceof Prototype) {
                Prototype prototype = (Prototype) item;
                if (prototype.getHoverVerification(xpos, ypos)) {
                    queue.add(prototype);
                } else {
                    prototype.setMouseHover(false);
                    if (item instanceof IFloating && action == InputEventType.MousePress) {
                        IFloating floatItem = (IFloating) item;
                        if (floatItem.isOutsideClickClosable()) {
                            if (item instanceof ContextMenu) {
                                ContextMenu cmToClose = (ContextMenu) item;
                                if (cmToClose.closeDependencies(margs)) {
                                    floatItem.hide();
                                }
                            } else {
                                floatItem.hide(margs);
                            }
                        }
                    }
                }
            }
        }

        if (queue.size() > 0) {
            if (hoveredItem != null && hoveredItem != queue.get(queue.size() - 1))
                manager.assignActionsForSender(InputEventType.MouseLeave, margs, hoveredItem, underFocusedItems,
                        false);

            hoveredItem = queue.get(queue.size() - 1);
            hoveredItem.setMouseHover(true);
            CommonService.currentCursor = hoveredItem.getCursor();
            glfw.SetCursor(handler.getWindowId(), CommonService.currentCursor.getCursor());

            if (window.isBorderHidden && window.isResizable && !window.isMaximized) {
                int handlerContainerWidth = rootContainer.getWidth();
                int handlerContainerHeight = rootContainer.getHeight();

                boolean cursorNearLeftTop = (xpos <= SpaceVILConstants.borderCursorTolerance
                        && ypos <= SpaceVILConstants.borderCursorTolerance);
                boolean cursorNearLeftBottom = (ypos >= handlerContainerHeight - SpaceVILConstants.borderCursorTolerance
                        && xpos <= SpaceVILConstants.borderCursorTolerance);
                boolean cursorNearRightTop = (xpos >= handlerContainerWidth - SpaceVILConstants.borderCursorTolerance
                        && ypos <= SpaceVILConstants.borderCursorTolerance);
                boolean cursorNearRightBottom = (xpos >= handlerContainerWidth - SpaceVILConstants.borderCursorTolerance
                        && ypos >= handlerContainerHeight - SpaceVILConstants.borderCursorTolerance);

                if (cursorNearRightTop || cursorNearRightBottom || cursorNearLeftBottom || cursorNearLeftTop) {
                    handler.setCursorType(EmbeddedCursor.Crosshair.getValue());
                } else {
                    if (xpos > handlerContainerWidth - SpaceVILConstants.borderCursorTolerance
                            || xpos <= SpaceVILConstants.borderCursorTolerance) {
                        handler.setCursorType(EmbeddedCursor.ResizeX.getValue());
                    }

                    if (ypos > handlerContainerHeight - SpaceVILConstants.borderCursorTolerance
                            || ypos <= SpaceVILConstants.borderCursorTolerance) {
                        handler.setCursorType(EmbeddedCursor.ResizeY.getValue());
                    }
                }
            }

            underHoveredItems = queue;
            Deque<Prototype> tmp = new ArrayDeque<>(underHoveredItems);
            while (!tmp.isEmpty()) {
                Prototype item = tmp.pollLast();
                if (item.equals(hoveredItem) && hoveredItem.isDisabled())
                    continue;
                item.setMouseHover(true);
                if (!item.isPassEvents(InputEventType.MouseHover))
                    break;
            }
            manager.assignActionsForHoveredItem(InputEventType.MouseHover, margs, hoveredItem, underHoveredItems,
                    false);
            return true;
        } else
            return false;
    }

    private List<IBaseItem> getInnerItems(Prototype root) {
        List<IBaseItem> list = new LinkedList<IBaseItem>();
        List<IBaseItem> root_items = root.getItems();

        for (IBaseItem item : root_items) {
            if (!item.isVisible() || !item.isDrawable())
                continue;
            if (item instanceof Prototype) {
                Prototype leaf = (Prototype) item;
                if (leaf.isDisabled())
                    continue;
                list.add(item);
                list.addAll(getInnerItems(leaf));
            }
        }
        return list;
    }

    private List<IBaseItem> getInnerItems(Prototype root, float xpos, float ypos) {
        List<IBaseItem> list = new LinkedList<IBaseItem>();
        List<IBaseItem> rootItems = root.getItems();

        for (IBaseItem item : rootItems) {
            if (!item.isVisible() || !item.isDrawable())
                continue;
            if (item instanceof Prototype) {
                Prototype leaf = (Prototype) item;
                if (leaf.isDisabled())
                    continue;
                if (leaf.getHoverVerification(xpos, ypos))
                    list.add(item);
                list.addAll(getInnerItems(leaf, xpos, ypos));
            }
        }
        return list;
    }

    <T> Prototype isInListHoveredItems(Class<T> type) {
        Prototype wanted = null;
        List<Prototype> list = new LinkedList<Prototype>(underHoveredItems);
        for (Prototype item : list) {
            try {
                boolean found = type.isInstance(item);
                if (found) {
                    wanted = item;
                    if (wanted instanceof IFloating)
                        return wanted;
                }
            } catch (ClassCastException cce) {
                continue;
            }
        }
        return wanted;
    }

    void findUnderFocusedItems(Prototype item) {
        Deque<Prototype> queue = new ArrayDeque<>();
        if (item == rootContainer) {
            underFocusedItems = null;
            return;
        }
        Prototype parent = item.getParent();
        while (parent != null) {
            queue.addFirst(parent);
            parent = parent.getParent();
        }
        underFocusedItems = new LinkedList<Prototype>(queue);
        underFocusedItems.remove(focusedItem);
    }

    void setFocusedItem(Prototype item) {
        if (item == null) {
            focusedItem = null;
            return;
        }
        if (focusedItem != null) {
            if (focusedItem.equals(item))
                return;
            focusedItem.setFocused(false);
        }
        focusedItem = item;
        focusedItem.setFocused(true);
        findUnderFocusedItems(item);
    }

    void resetFocus() {
        if (focusedItem != null)
            focusedItem.setFocused(false);
        focusedItem = rootContainer;
        focusedItem.setFocused(true);
        if (underFocusedItems != null)
            underFocusedItems.clear();
    }

    void resetItems() {
        resetFocus();
        if (hoveredItem != null)
            hoveredItem.setMouseHover(false);
        hoveredItem = null;
        underHoveredItems.clear();
    }

    static String getResourceString(String resource) {
        StringBuilder result = new StringBuilder();
        InputStream inputStream = CommonProcessor.class.getResourceAsStream(resource);
        try (BufferedReader scanner = new BufferedReader(new InputStreamReader(inputStream))) {
            String line;
            while ((line = scanner.readLine()) != null) {
                result.append(line).append("\n");
            }

        } catch (IOException e) {
            e.printStackTrace();
        }
        return result.toString();
    }
}